lib/modifications.rb

```diff
@@ -504,5 +662,63 @@ def self.create_bundle_entry(resource)
         entry.request.url = resource.resourceType
         entry
     end
+
+    def self.get_reference_type(b, reference_string)
+      id = reference_string.split(':').last
+      b.entry.find { |e| e.resource.id == id }&.resource&.class
+    end
+
+    def self.get_resource_counts(bundle)
+      resource_counts = {}
+      bundle.entry.each do |entry|
+        resource_type = entry.resource.resourceType
+        if resource_counts[resource_type]
+          resource_counts[resource_type] += 1
+        else
+          resource_counts[resource_type] = 1
+        end
+      end
+      resource_counts
+    end
+
+    def self.get_diagreport_referenced_observations(bundle)
+      bundle.entry.flat_map do |entry|
+        next unless entry.resource.is_a? FHIR::DiagnosticReport
+        entry.resource.result.map { |r| r&.reference&.split(':')&.last }
+      end.uniq
+    end
+
+    def self.get_docref_referenced_attachments(bundle)
+      docrefs = bundle.entry.find_all { |e| e.resource.is_a? FHIR::DocumentReference}
+      docrefs.map do |docref|
+        bundle.entry.reverse.find { |e|
+          e&.resource&.resourceType == 'DiagnosticReport' &&
+            e&.resource&.presentedForm&.first&.data == docref&.resource&.content&.first&.attachment&.data }&.resource&.id
```

Comment:
Do you need to worry about `nil == nil` here with all of the safe navigation?

Reply:
Yeah, probably. I'll add some checking in here.

Reply:
Like above...probably. Good call, I'll add a check.